src/font2swf.1
.TH font2swf "1" "April 2012" "font2swf" "swftools"
.SH NAME
font2swf \- Converts a font to SWF.

.SH Synopsis
.B font2swf <fontfile>

.SH DESCRIPTION
Takes a font file (.ttf, .afm, .pfa, .pfb and all other types
supported by FreeType) and converts it into a SWF file.
.PP
The SWF will contain the Font in SWF format (that is, a
DefineFont2 Tag) as well as a textfield containing all the
characters the font has. This means the resulting SWF will
be viewable.

.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR 
    Print short help message and exit
.TP
\fB\-v\fR, \fB\-\-verbose\fR 
    Be verbose. Use more than one -v for greater effect.
.TP
\fB\-o\fR, \fB\-\-output\fR \fIfilename\fR
    Write output to file \fIfilename\fR. (If not given, the output will go
    to a file called "output.swf")
.TP
\fB\-V\fR, \fB\-\-version\fR 
    Print version info and exit
.SH AUTHOR

Matthias Kramm <kramm@quiss.org>


ChangeLog
0.0.1: initial release
0.0.2: 
	* pdf2swf: Added support for pdf bitmaps, clipping, state saving/restoring, colors,
	           solid fill and line width. The output SWF files are now much smaller.
	* swfcombine: Added stacking option (-t)
	* swfcombine: Small bugfixes with the command line options
	* rfxswf-lib: some bugfixes concerning sprites, cgi, shapes, jpegs.
	* swfstrings: New: Allows extraction of text from swf files
	* swfdump: New: Dumps swf file information. (And does some error checking on the SWF)
	* jpeg2swf: New: Converts one or more jpegs to a swf movie.

0.1.0: 
	* configure: Now works without automake, automatically disables 
			jpeg2swf and pdf2swf if jpeglib is not found (disable
			both) or t1lib is not found (disable pdf2swf)
	* swfcombine: bugfix: DEFINETEXT tags didn't get relocated right
	* swfcombine: bugfix: no longer builds "unstreamable" SWFs
	* rfxswf-lib: actionscript
	* rfxswf-lib: lossless image formats
	* pdf2swf: conversion of lossless image formats
	* pdf2swf: picture cache to prevent images getting stored more than once
	* pdf2swf: conversion of (hyper)links
	* swfdump: -a, -e and -XY options
	* lib/example/zlibtest: new file, demonstrates using lossless images with rfxswf-lib
	* lib/example/dumpfont: utility for dumping fonts out of swf (and generate c sourcecode for creating them)
	* lib/example/demofont: example font, generated with lib/example/dumpfont.

0.1.1:  
	* swfcombine: Wasn't able to handle converted pdfs with links. Fixed.
	* swfcombine: Introduce --dummy, -X,-Y,-r options
	* swfdump: new: -r option
	* pdf2swf: small bugfixes
	* HOWTO_pdf2swf: added section about correcting size and framerate

0.1.2:
	* swfdump: can dump all actionscript flavours (button, button2, standalone) now
	* rfxswf-lib: improved font handling
	* lib/example/makefonts.cc: new file, converts the standard fonts to swf
	* lib/example/glyphshape.c: new file, demonstrates using font characters as shapes
	* pdfswf: many bugfixes. The "pdfscreen" manual now get's converted correctly
	* rfxswf-lib: bugfix: Couldn't handle jpeg grayscale images

0.2.0:	
	* pdf2swf: TrueType support added. (using ttf2pt1)
	* rfxswf-lib: performs conversion of jpeg colorspaces now. (Makes pdf2swf cope
	               with more types of jpegs)
	* swfextract: new tool, allows to extract single objects, frames etc. from movieclips
	* swfcombine: allows ids instead of names, now, too.
	* swfdump: added text displaying functionality. (like swfstrings)
	* swfdump: now dumps also the undocumented placeobject2 actioncode.
	* lib/example/avi2swf.cc: For now, converts small (!) avi movie files to swf.  (using 
			          libavifile)

0.2.1:	
	* configure: Does checking for byte order and command substitution capability now.
	* rfxswf: Applied Anjo Krank's patch for MacOS X.
	* pdf2swf: Fixed font transformation matrix
	* swfcombine: new options: --cat, --stack1, --merge, --frame
	* swfs/: new directory, holds some standard SWFs primarily used for pdf2swf
	         installs into ${DATADIR}/swftools/swfs
	* pdf2swf: New options --viewer, --preloader, -b, -l
	* pdf2swf: Fonts are now in ${DATADIR}/swftools/fonts

0.2.2:	
	* swfs/: Added/Wrote sources for some of the swf files 
	* swfdump: Major bugfix in placeobject2, hex output function
	* swfextract: Can now extract png, jpeg and mp3.
	* swfcombine: big endian fixes
	* swfextract: big endian fixes
	* pdf2swf: Changed from 100dpi to 72dpi conversion
	* pdf2swf: Some optimizations, fix the zero string problem.
	* lib/: Added Bladeenc 0.94.1 for mp3 compression
	* lib/example/avi2swf.cc: Can now handle big files. Slightly better compression ratio. Sound.
	* lib/example/sound.c: New file, demonstrates use of the sound module in librfxswf.

0.2.3: 
	* pdf2swf: New option --fontdir, can now handle external fonts.
	* lib/modules/swftext.c: Several font-extraction bugfixes
	* src/swfdump: Bugfixes, ids and depths are now in decimal
	* src/swfcombine: Bugfixes.
	* configure.in: Cleaned up.

0.3.0:
	* pdf2swf: The --ignore option now actually makes the pdf files smaller.
	* rfxswf: Can now write DefineFont2 tags
	* swfextract/swfcombine: Bugfixes (DefineEditText)
	* swfdump: New option --used, prints out referred IDs of each tag.
	* rfxswf: Can now read and write Flash 6 (MX) compressed SWFs. (SWCs)
	* pdf2swf: Added compression flag (-z)
	* swfcombine: Added compression flag (-z)
	* configure.in: zlib is now required.
	* src/png2swf: New tool.
	* swfdump: Prints some image infos now.
	* pdf2swf: Fixed Image handling
	* avi2swf/: Is part of the main distribution now
	* src/wav2swf: New tool.

0.3.1:
	* configure: Didn't create all Makefiles when t1lib or jpeglib were absent. Fixed.
	* swfs/Makefile: FreeBSD fixes.

0.4.0:
	* pdf2swf/xpdf: Upgraded PDF Parser to Version 1.01.
	* pdf2swf: Can now handle Type 3 Fonts. (As Bitmaps)
	* src/swfcombine: Completely rewritten, using the rfxswf routines
	* configure: Check for libtool
	* lib/rfxswf: Fixed some parsing bugs (swffont, swfdump)
	* pdf2swf: Added antialiasing capabilities for Type 3 Fonts
	* lib/examples/makefonts: Fonts: Bounding boxes are now handled correctly
	* png2swf: Support for more image types
	* pdf2swf: New option -t (--stop)

0.4.1:
	* src/swfcombine: Bugfix with -x option
	* lib/Makefile: MacOSX fixes
	* pdf2swf/SWFOutput.cc: fixed undefined sqrt()
	* src/png2swf: Support for transparency
	* pdf2swf: Support for CID Fonts
	* src/swfcombine: Bugfixes
	* lib/rfxswf: Added Sprite Unfolding routines

0.4.2:
	* src/Makefile: OpenBSD fixes
	* lib/lame: Removed *.o files from archive
	* avi2swf: Implemented --flip option
	* pdf2swf: Implemented --flashversion option
	* pdf2swf: Implemented Viewer gadgets- it's now possible to call Flash functions via hyperlink

0.4.3:
	* src/jpeg2swf: Grayscale fix
	* lib/rfxswf: Improved MX support
	* pdf2swf: Command line fixes
	* swfdump: Added -f (--frames) option
	* png2swf: Added support for mode 6 images
	* swfs: Support for print protection in simple_viewer

src/font2swf.c
/* font2swf.c

   Utility for converting TrueType and Type 1 fonts to SWF.
   
   Part of the swftools package.

   Copyright (c) 2004 Matthias Kramm <kramm@quiss.org>
 
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */

#include <stdio.h>
#include <unistd.h>
#include <math.h>
#include "../lib/rfxswf.h"
#include "../lib/args.h"

static char * filename = 0;
static char * destfilename = "output.swf";
static int all=0;
static int verbose=0;
static char * fontname = 0;
static char config_flashtype = 0;

static struct options_t options[] = {
{"h", "help"},
{"v", "verbose"},
{"o", "output"},
{"V", "version"},
{0,0}
};

int args_callback_option(char*name,char*val)
{
    if(!strcmp(name, "V")) {
        printf("font2swf - part of %s %s\n", PACKAGE, VERSION);
        exit(0);
    }
    else if(!strcmp(name, "o")) {
	destfilename = val;
	return 1;
    }
    else if(!strcmp(name, "v")) {
	verbose ++;
	return 0;
    }
    else if(!strcmp(name, "T")) {
	config_flashtype=1;
	return 0;
    }
    else if(!strcmp(name, "n")) {
	fontname = val;
	return 1;
    }
    else if(!strcmp(name, "a")) {
	all = 1;
	return 0;
    }
    else {
        printf("Unknown option: -%s\n", name);
	exit(1);
    }
    return 0;
}
int args_callback_longoption(char*name,char*val)
{
    return args_long2shortoption(options, name, val);
}
void args_callback_usage(char *name)
{
    printf("\n");
    printf("Usage: %s <fontfile>\n", name);
    printf("\n");
    printf("-h , --help                    Print short help message and exit\n");
    printf("-v , --verbose                 Be verbose. Use more than one -v for greater effect.\n");
    printf("-o , --output <filename>       Write output to file <filename>.\n");
    printf("-V , --version                 Print version info and exit\n");
    printf("\n");
}
int args_callback_command(char*name,char*val)
{
    if(filename) {
        fprintf(stderr, "Please specify only one font\n");
        exit(1);
    }
    filename = name;
    return 0;
}

static void convertFont(char*infile, char*outfile)
{
    SWFFONT * font;
    
    font = swf_LoadFont(infile, config_flashtype);
    swf_FontCreateAlignZones(font);

    if(fontname)
        font->name = strdup(fontname);

    swf_WriteFont(font, outfile);
    swf_FontFree(font);
}

int main(int argc, char ** argv)
{
    char cwd[128];
    getcwd(cwd, 128);
    processargs(argc, argv);
    if(!all && !filename) {
	fprintf(stderr, "You must supply a filename.\n");
	exit(1);
    }
   
    if(!all) {
	convertFont(filename, destfilename);
	return 0;
    }

    /* TODO */
    printf("--all not implemented yet.\n");

    return 0;
}


